File: packages/core/src/system/editor.ts
import * as client from './client';
import {
  EditorInterface,
  SystemConfig,
  ViewFileParams,
  CreateFileParams,
  ReplaceStringParams,
  InsertTextParams,
  UndoEditParams,
} from './types';

/**
 * Editor control implementation
 * Provides functionality for file manipulation operations
 */
export class Editor implements EditorInterface {
  constructor(private config: SystemConfig) {}

  /**
   * View contents of a file
   * @param params File viewing parameters
   * @returns Promise with file contents
   * @throws Error if file viewing fails
   */
  async viewFile({ path, viewRange }: ViewFileParams): Promise<string> {
    const response = await client.viewFile({
      body: {
        path,
        view_range: viewRange ?? null,
      },
    });

    if (!response.data) {
      throw new Error('Failed to view file');
    }

    return response.data as string;
  }

  /**
   * Create a new file with optional content
   * @param params File creation parameters
   * @throws Error if file creation fails
   */
  async createFile({ path, content }: CreateFileParams): Promise<void> {
    await client.createFile({
      body: {
        path,
        file_text: content ?? null,
      },
    });
  }

  /**
   * Replace string occurrences in a file
   * @param params String replacement parameters
   * @throws Error if string replacement fails
   */
  async replaceString({
    path,
    oldStr,
    newStr,
  }: ReplaceStringParams): Promise<void> {
    await client.replaceString({
      body: { path, old_str: oldStr, new_str: newStr },
    });
  }

  /**
   * Insert text at a specific line in a file
   * @param params Text insertion parameters
   * @throws Error if text insertion fails
   */
  async insertText({ path, line, text }: InsertTextParams): Promise<void> {
    await client.insertText({
      body: { path, insert_line: line, text },
    });
  }

  /**
   * Undo the last edit operation
   * @param params Undo parameters
   * @throws Error if undo operation fails
   */
  async undoLastEdit({ path }: UndoEditParams): Promise<void> {
    await client.undoEdit({
      body: { path },
    });
  }
}

File: platforms/linux/api/services/editor.py
import os
from typing import List, Optional, Tuple

class EditorService:
    def __init__(self):
        self._file_history: dict[str, List[str]] = {}  # For undo functionality

    async def view_file(self, path: str, view_range: Optional[Tuple[int, int]] = None) -> str:
        """
        View contents of a file, optionally within a specific line range.
        """
        if not os.path.exists(path):
            raise FileNotFoundError(f"File not found: {path}")

        with open(path, 'r') as file:
            lines = file.readlines()

        if view_range:
            start, end = view_range
            start = max(0, start - 1)  # Convert to 0-based index
            end = min(len(lines), end)
            lines = lines[start:end]

        return "".join(lines)

    async def create_file(self, path: str, content: Optional[str] = None) -> None:
        """
        Create a new file with optional content.
        """
        if os.path.exists(path):
            raise FileExistsError(f"File already exists: {path}")

        # Create directory if it doesn't exist
        os.makedirs(os.path.dirname(path), exist_ok=True)

        with open(path, 'w') as file:
            if content:
                file.write(content)

        # Initialize history for the file
        self._file_history[path] = []

    async def replace_string(self, path: str, old_str: str, new_str: str) -> None:
        """
        Replace all occurrences of a string in a file.
        """
        if not os.path.exists(path):
            raise FileNotFoundError(f"File not found: {path}")

        # Read current content
        with open(path, 'r') as file:
            content = file.read()

        # Save current content to history
        if path not in self._file_history:
            self._file_history[path] = []
        self._file_history[path].append(content)

        # Replace string and write back
        new_content = content.replace(old_str, new_str)
        with open(path, 'w') as file:
            file.write(new_content)

    async def insert_text(self, path: str, text: str, insert_line: Optional[int] = None) -> None:
        """
        Insert text at a specific line in the file.
        If insert_line is None, append to the end of the file.
        """
        if not os.path.exists(path):
            raise FileNotFoundError(f"File not found: {path}")

        # Read current content
        with open(path, 'r') as file:
            lines = file.readlines()

        # Save current content to history
        if path not in self._file_history:
            self._file_history[path] = []
        self._file_history[path].append("".join(lines))

        # Prepare text for insertion
        if not text.endswith('\n'):
            text += '\n'

        # Insert text at specified line or append
        if insert_line is None:
            insert_line = len(lines)
        else:
            insert_line = max(0, min(len(lines), insert_line - 1))  # Convert to 0-based index
        
        lines.insert(insert_line, text)

        # Write back to file
        with open(path, 'w') as file:
            file.writelines(lines)

    async def undo_last_edit(self, path: str) -> None:
        """
        Undo the last edit made to a file.
        """
        if not os.path.exists(path):
            raise FileNotFoundError(f"File not found: {path}")

        if path not in self._file_history or not self._file_history[path]:
            raise ValueError(f"No history available for file: {path}")

        # Get the last version and remove it from history
        previous_content = self._file_history[path].pop()

        # Write the previous version back to file
        with open(path, 'w') as file:
            file.write(previous_content) 